### PR TITLE
Request types in separate services with validation

### LINKED ISSUE #5: Encapsulate each type of request in separate service with typization
Encapsulate each type of supported request. Add typization of all additional fields in request type. Types mapping should be defined in application configuration. Any unknown fields should be treated as simple strings.

Relates #3

### PR DESCRIPTION
**Pull request by laska-sl**

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -2,16 +2,14 @@
 [*]
 
 # ReSharper properties
-resharper_blank_lines_after_start_comment=0
```
Comment (sponkrashin) [inline .editorconfig:4]: Why did you remove some rules? They seem useful.

File: ArcadiaTeamsBot/appsettings.json [hunk 2]
```diff
@@ -10,5 +10,10 @@
   "ServiceDesk": {
     "ApiUrl": "",
     "ApiKey": ""
+  },
+  "ServiceDeskRequestsMapping": {
+    "Second field": "Number",
```
Comment (sponkrashin) [inline ArcadiaTeamsBot/appsettings.json]: It should be request type specific, not global.

File: ArcadiaTeamsBot/Startup.cs [hunk 3]
```diff
@@ -36,6 +40,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddTransient<IBot, Bot>();
             services.AddSingleton<IBotFrameworkHttpAdapter, BotAdapterWithErrorHandling>();
             services.AddSingleton(this.serviceDeskConfiguration);
+            services.AddSingleton(this.serviceDeskMappingConfiguration);
```
Comment (sponkrashin) [inline ArcadiaTeamsBot/Startup.cs]: Not the best way to register in DI container such common type as Dictionary. It will work, but violates some SOLID rules.

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeForUIField.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿namespace ArcadiaTeamsBot.ServiceDesk.Requests
+{
+    public class RequestTypeForUIField
+    {
+        public RequestTypeForUIField(string fieldName, RequestTypeForUIFieldType fieldType)
+        {
+            this.FieldName = fieldName;
+            this.FieldType = fieldType;
+        }
+        public string FieldName { get; set; }
```
Comment (sponkrashin) [inline ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeForUIField.cs]: ...

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeForUI.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+        public RequestTypeForUI(Dictionary<string, string> serviceDeskMappingConfiguration)
+        {
+            this.serviceDeskMappingConfiguration = serviceDeskMappingConfiguration;
+            this.RequestTypeForUIFields = new List<RequestTypeForUIField>();
+        }
+
+        public List<RequestTypeForUIField> RequestTypeForUIFields { get; set; }
+
+        public void LoadFieldsTypes(ServiceDeskRequestTypeDTO requestTypeDTO)
```
Comment (sponkrashin) [inline ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeForUI.cs]: Instead all of this it would better be a factory which will manage dependencies and create types for UI by specified request type DTO.

**Review by sponkrashin — CHANGES_REQUESTED** [review]
By the way, where is any kind of validation?

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeForUIField.cs (hunk 4 above)
Comment (laska-sl) [reply ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeForUIField.cs]: Sorry, forgot to autoformat.
Already fixed.

File: ArcadiaTeamsBot/Startup.cs (hunk 3 above)
Comment (laska-sl) [reply ArcadiaTeamsBot/Startup.cs]: Made it separate model with nested Dictionary.

File: ArcadiaTeamsBot/appsettings.json (hunk 2 above)
Comment (laska-sl) [reply ArcadiaTeamsBot/appsettings.json]: Changed appsettings.json structure.

File: .editorconfig (hunk 1 above)
Comment (laska-sl) [reply .editorconfig:4]: Returned some rules.

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeForUI.cs (hunk 5 above)
Comment (laska-sl) [reply ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeForUI.cs]: Implemented factory method

File: .editorconfig [hunk 6]
```diff
@@ -3,14 +3,19 @@
...
+resharper_braces_for_for=required
 resharper_braces_for_foreach=required
+resharper_braces_for_ifelse=required
+resharper_braces_for_while=required
 resharper_braces_redundant=true
 resharper_csharp_insert_final_newline=true
+resharper_csharp_keep_blank_lines_in_code=1
 resharper_csharp_keep_blank_lines_in_declarations=1
 resharper_csharp_max_line_length=200
-resharper_place_expr_method_on_single_line=false
```
Comment (sponkrashin) [inline .editorconfig:12]: Why?

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestType/IRequestTypeForUI.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿namespace ArcadiaTeamsBot.ServiceDesk.Requests.RequestType
+{
+    using System.Collections.Generic;
+
+    public interface IRequestTypeForUI
+    {
+        List<RequestTypeForUIField> RequestTypeForUIFields { get; }
```
Comment (sponkrashin) [inline ArcadiaTeamsBot.ServiceDesk.Requests/RequestType/IRequestTypeForUI.cs]: Don't expose concrete types, especially such type as `List`. It should be of type `IEnumerable<...>`.

Comment (sponkrashin) [reply ArcadiaTeamsBot.ServiceDesk.Requests/RequestType/IRequestTypeForUI.cs]: If you expose from interface type which state can be changed, it may lead to some unexpected errors. Imagine:
```cs
interface IA
{
    List<string> Items { get; }
}

class A : IA
{
    A()
    {
        this.Items = new List<string>();
        this.Items.Add("some mandatory value which should always exist");
    }

    List<string> Items { get; }
}

var a = new A();
a.Items.Clear();
```

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypesMappingConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+﻿namespace ArcadiaTeamsBot.ServiceDesk.Requests
+{
+    using System.Collections.Generic;
+
+    public class RequestTypesMappingConfiguration : Dictionary<string, Dictionary<string, string>>
```
Comment (sponkrashin) [inline ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypesMappingConfiguration.cs]: You can specify that values in inner dictionary will be of type `RequestTypeForUIFieldType`.

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeFactory/RequestTypeForUIFactory.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            var requestTypeForUIFields = new List<RequestTypeForUIField>();
+
+            this.requestTypesMappingConfiguration.TryGetValue(requestTypeDTO.Title, out var requestTypeFromConfig);
+
+            foreach (var fieldFromDTO in requestTypeDTO.RequestTypeFields)
+            {
+                var fieldName = fieldFromDTO.FieldName;
+
+                if (requestTypeFromConfig == null ||
```
Comment (sponkrashin) [inline ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeFactory/RequestTypeForUIFactory.cs]: Use null-propagation operator and get rid of Enum.TryParse by changing settings model.

**Review by sponkrashin — CHANGES_REQUESTED** [review]

**Review by sponkrashin — CHANGES_REQUESTED** [review]
Once again - where is validation?

File: .editorconfig (hunk 6 above)
Comment (laska-sl) [reply .editorconfig:12]: There are 3 options:
image

1. Never = "false"
image

2. Always = "true"
image

3. If owner is singleline = nothing
image

"If owner is singleline" looks more appropriate.

Comment (sponkrashin) [reply .editorconfig:12]: Then why it was false before?

Comment (laska-sl) [reply .editorconfig:12]: Was by default or added by mistake during first R# setup.

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypesMappingConfiguration.cs (hunk 8 above)
Comment (laska-sl) [reply ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypesMappingConfiguration.cs]: Sure. Fixed

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestType/IRequestTypeForUI.cs (hunk 7 above)
Comment (laska-sl) [reply ArcadiaTeamsBot.ServiceDesk.Requests/RequestType/IRequestTypeForUI.cs]: Fixed

File: ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeFactory/RequestTypeForUIFactory.cs (hunk 9 above)
Comment (laska-sl) [reply ArcadiaTeamsBot.ServiceDesk.Requests/RequestTypeFactory/RequestTypeForUIFactory.cs]: I've changed settings model and condition structure.

**Review by sponkrashin — CHANGES_REQUESTED** [review]
Could you, please, rename all new classes and variables such that they doesn't contain prepositions? They just pollute the code without any special meaning.

**Comment by laska-sl** [comment]
> Could you, please, rename all new classes and variables such that they doesn't contain prepositions? They just pollute the code without any special meaning.

File: ArcadiaTeamsBot.CQRS.Abstractions/Queries/GetCurrentRequestsQuery.cs [hunk 10]
```diff
@@ -6,9 +6,9 @@
 
     using MediatR;
 
-    public class GetCurrentServiceDeskRequestsQuery : IRequest<IEnumerable<ServiceDeskRequestDTO>>
+    public class GetCurrentRequestsQuery : IRequest<IEnumerable<RequestDTO>>
```
Comment (sponkrashin) [inline ArcadiaTeamsBot.CQRS.Abstractions/Queries/GetCurrentRequestsQuery.cs]: `ServiceDesk` is not a preposition. It was better before.

**Review by sponkrashin — CHANGES_REQUESTED** [review]

File: ArcadiaTeamsBot.CQRS.Abstractions/Queries/GetCurrentRequestsQuery.cs (hunk 10 above)
Comment (laska-sl) [reply ArcadiaTeamsBot.CQRS.Abstractions/Queries/GetCurrentRequestsQuery.cs]: I know, but preferred to remove this from names, because it's obvious from project context.

Comment (laska-sl) [reply ArcadiaTeamsBot.CQRS.Abstractions/Queries/GetCurrentRequestsQuery.cs]: If it was better, I've returned names

**Review by sponkrashin — APPROVED** [review]
